[Merged by Bors] - Set max size for variable size fields in types encoded with scale

fasmat opened the pull request:
## Motivation
Closes https://github.com/spacemeshos/go-spacemesh/issues/4095

## Changes
Use newest go-scale that fails code generation if the "max" tag isn't set for variable size fields in types encoded with go-scale and fix code generation errors caused by missing tags.

- Some types had to be restructured (specifically named) to make tagging possible
- Some types were changed from slices to arrays, since their size is constant
- Here are all the types I changed, some of which I am unsure about. Feedback is appreciated:

Test
====

| Type | Field name | Max. Size | Description |
| ----- | ----------- | --------- | ------------ |
| `beacon.Proposal` | - | `[4]bytes` | Used to be just `[]byte`, since a proposal is always 4 bytes; converted to an array and added `DecodeScale` / `EncodeScale` methods. Additionally added a `String` method and a `ProposalFromString` function since these are often converted from and to strings (usually to be used as key in a `map`) |
| `beacon.ProposalMessage` | `VRFSignature` | 80 | VRF Signature of size 80 |
| `beacon.FirstVotingMessageBody` | `ValidProposals` | 1000 | The number of valid proposals should be lower than 100, max size of 1000 as a safety limit |
| `beacon.FirstVotingMessageBody` | `PotentiallyValidProposals` | 1000 | The number of valid proposals should be lower than 100, max size of 1000 as a safety limit |
| `beacon.FirstVotingMessage` | `Signature` | 64 | ED25519 Signature of size 64 |
| `beacon.FollowingVotingMessageBody` | `VotesBitVector` | 128 | We expect less than 1000 votes -> 1000/8 = 125 bytes; 128 bytes should be sufficient for encoding |
| `beacon.FollowingVotingMessage` | `Signature` | 64 | ED25519 Signature of size 64 |
| `weakcoin.Message` | `MinerPK` | 32 | ED25519 Public key of size 32 |
| | `VrfSignature` | 80 | VRF Signature of size 80 |
| `type.Account` | `State` | 10_000 | ??? (m: Template for State should be much smaller than 10k ?) |
| `type.NIPostChallenge` | `InitialPostIndices` | 8000 | Should be K2 * 8, some buffer for safety |
| `type.ActivationTx` | `Signature` | 64 | ED25519 Signature of size 64 |
| `type.PoetProof` | `Members` | 100_000 | Depends on the number of Smeshers we expect committing a challenge to a PoET |
| `type.PoetProofMessage` | `PoetServiceID` | 32 | ED25519 Public key for the PoetService |
| | `RoundID` | 32 | Identifier for a poet round. string representation of round count since genesis |
| | `Signature` | 64 | ED25519 Signature of size 64 |
| `type.PoetRound` | `ID` | 32 | Identifier for a poet round. string representation of round count since genesis |
| `type.Post` | `Indices` | 8000 | Should be K2 * 8, some buffer for safety |
| `type.PostMetadata` | `Challenge` | 32 | Poet challenge of size 32 byte |
| `type.ProcessingError` | `Err` | n/a | **TODO(mafa)** replace with Error codes |
| `type.Ballot` | `Signature` | 64 | ED25519 Signature of size 64 |
| | `EligibilityProofs` | 500 | number of expected proposals per layer is 50, 10X for safety margin |
| `type.InnerBallot` | `Support` | 10_000 | tortoise sliding window 10K layers |
| | `Against` | 10_000 | tortoise sliding window 10K layers |
| | `Abstain` | 10_000 | tortoise sliding window 10K layers |
| `type.EpochData` | `ActiveSet` | 100_000 | Depends on the number of Smeshers we expect publishing an ATX in an epoch |
| `type.InnerBlock` | `Rewards` | 500 | number of expected proposals per layer is 50, 10X for safety margin |
| | `TxIDs` | 100_000 | The upper limit (with buffer) on the number of transactions we expect in a block |
| `type.Certificate` | `Signatures` | 1000 | Depends on the Hare Parameter `N` + some buffer |
| `type.CertifyMessage` | `Signature` | 64 | ED25519 Signature of size 64 |
| `type.CertifyContent` | `Proof` | 80 | VRF Signature of size 80 |
| `type.HareEligibilityGossip` | `PubKey` | 32 | ED25519 Public key of size 32 |
| `type.HareEligibility` | `Proof` | 80 | VRF Signature of size 80 |
| `type.VotingEligibility` | `Sig` | 80 | VRF Signature of size 80 |
| `type.AtxProofMsg` | `Signature` | 64 | ED25519 Signature of size 64 |
| `type.BallotProofMsg` | `Signature` | 64 | ED25519 Signature of size 64 |
| `type.HareProofMsg` | `Signature` | 64 | ED25519 Signature of size 64 |
| `type.PoetServiceID` | `ServiceID` | 32 | ED25519 Public key for the PoetService |
| `type.NiPostBuilderState` | `PoetRequests` | 100 | 100 should be a safe limit for number of PoETs used by a single node |
| `type.PoetProofRef` | - | `[32]byte` | Used to be just `[]byte`, since always 32 bytes converted to an array |
| `type.Proposal` | `Signature` | 64 | ED25519 Signature of size 64 |
| `type.InnerProposal` | `TxIDs` | 100_000 | Depends on `NumTXsPerProposal`, which is defined by the app config `TxsPerProposal` with a lot of extra margin |
| `type.TransactionResult` | `Message` | 1024 | human readable error. **TODO(mafa)** convert from string to enum |
| | `Addresses` | 10 | Depending on type of transaction 1-3 with some safety buffer |
| `type.RawTx` | `Raw` | 4000 | A transaction is expected to be less than 4 KiB in size |
| `fetch.RequestMessage` | `Hint` | ? (currently 256) | **TODO(mafa)** convert from string to enum |
| `fetch.ResponseMessage` | `Data` | 3_200_000 | 100K of ATXIDs |
| `fetch.RequestBatch` | `Requests` | 1000 | configurable via BatchSize |
| `fetch.ResponseBatch` | `Responses` | 1000 | configurable via BatchSize |
| `fetch.MeshHashes` | `Layers` | 1000 | configurable via `MaxHashesInReq` |
| | `Hashes` | 1000 | configurable via `MaxHashesInReq` |
| `fetch.MaliciousIDs` | `NodeIDs` | 100_000 | Max. number of malicious nodes we expect in the network |
| `fetch.EpochData` | `AtxIDs` | 100_000) | max number of atx in a epoch |
| `fetch.LayerData` | `Ballots` | 500 | number of expected ballots per layer is 50, 10X for safety margin |
| | `Blocks` | 100 | expected 1 per layer |
| `multisig.MultiSig` | `PublicKeys` | 10 | hardcoded for genesis VM |
| `multisig.SpawnArguments` | `PublicKeys` | 10 | hardcoded for genesis VM |
| `hare.Message` | `Signature` | 64 | ED25519 Signature of size 64 |
| `hare.Certificate` | `Values` | 500 | number of expected proposals per layer is 50, 10X for safety margin |
| `hare.AggregatedMessages` | `Messages` | 1000 | hare param N plus safety margin |
| `hare.InnerMessage` | `Values` | 1000 | hare param N plus safety margin |
| `handshake.HandshakeAck` | `Error` | n/a | **TODO(mafa)** replace with Error codes |
| `server.Response` | `Data` | 10485760 (= 10 MiB) | Generic server response type |
| | `Error` | n/a | **TODO(mafa)** replace with Error codes |

## Test Plan
n/a

## TODO
<!-- This section should be removed when all items are complete -->
- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed

## DevOps Notes
<!-- Please uncheck these items as applicable to make DevOps aware of changes that may affect releases -->
- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

fasmat committed:
Use newest go-scale

fasmat committed:
Add some max sizes to scale encoded types

fasmat committed:
go generate

fasmat committed:
Update go-scale to 1.1.5

fasmat committed:
Add scale max tags to Handshake types

fasmat committed:
Update poet

fasmat committed:
Update types in beacon package

fasmat committed:
WiP

fasmat committed:
Merge remote-tracking branch 'origin/develop' into 4095-enforce-max-size-go-scale

fasmat committed:
Remove more util functions

fasmat committed:
WiP

fasmat committed:
Merge remote-tracking branch 'origin/develop' into 4095-enforce-max-size-go-scale

fasmat committed:
Merge remote-tracking branch 'origin/develop' into 4095-enforce-max-size-go-scale

fasmat committed:
Merge remote-tracking branch 'origin/develop' into 4095-enforce-max-size-go-scale

fasmat committed:
Merge remote-tracking branch 'origin/develop' into 4095-enforce-max-size-go-scale

fasmat committed:
Add more size tags to scale encoded types

fasmat committed:
Add more size tags

fasmat committed:
Add more size tags

fasmat committed:
Use newest version of scale

fasmat committed:
Update poet

fasmat committed:
Merge remote-tracking branch 'origin/develop' into 4095-enforce-max-size-go-scale

fasmat committed:
cleanup

dshulyak commented on beacon/message.go:
```diff
@@ -17,31 +17,35 @@ type ProposalVrfMessage struct {
 type FirstVotingMessage struct {
 	FirstVotingMessageBody
-	Signature []byte
+	Signature []byte `scale:"max=32"`
```
i think signature is 64 bytes

dshulyak commented on beacon/message.go:
```diff
@@ -17,31 +17,35 @@ type ProposalVrfMessage struct {
-	ValidProposals            [][]byte
-	PotentiallyValidProposals [][]byte
+	ValidProposals            []Proposal `scale:"max=32"` // TODO(mafa): check if max. value is correct
+	PotentiallyValidProposals []Proposal `scale:"max=32"` // TODO(mafa): check if max. value is correct
```
maybe it makes sense to change Proposal to byte array (e.g [32]byte ) ? it is more compact, easier to decode and safer

@countvonzero

fasmat replied:
In some places (like signatures) this might work, but the size of `ValidProposals` and `PotentiallyValidProposals` is afaik a configuration parameter.

Arrays cannot be dynamic in size, that being said the `scale:"max=xxx"` tag also doesn't consider configuration...

countvonzero replied:
i think @dshulyak  is talking about the element `Proposal` itself. it is always 4 byte (defined as types.BeaconSize). so you can safely define it as a byte array

dshulyak replied:
yes i was talking about that

we need to set some upper bound on the total number of proposals as well, maybe 10000 is ok ?

countvonzero replied:
here the total number of proposals, with high probabilities, would be lower than 100. so i think 200 is enough.

fasmat replied:
Made Proposal a `[4]byte` instead of `[]byte` and added a max of 200 to `ValidProposals` and `PotentiallyValidProposals`

dshulyak commented on beacon/weakcoin/weak_coin.go:
```diff
@@ -41,8 +41,8 @@ type Message struct {
-	MinerPK      []byte
-	VrfSignature []byte
+	MinerPK      []byte `scale:"max=32"`
+	VrfSignature []byte `scale:"max=32"`
```
vrf signature is 80 bytes

fasmat committed:
Signatures should be 64 bytes

poszu commented on activation/poet.go:
```diff
@@ -118,41 +118,64 @@ func (c *HTTPPoetClient) Submit(ctx context.Context, challenge []byte, signature
 
+	p := resBody.Proof.GetProof()
+	root := p.GetRoot()
+	leaves := make([]shared.Leaf, 0, len(p.ProvenLeaves))
```
The original code used getters like `p.GetProvenLeaves()`. The reason is that in protobuf every field is _optional_ and the getters have nil-checks. It is possible that both `err` and `p` are `nil` here - the data from the poet cannot be trusted. Direct access can lead to a segfault.

fasmat replied:
Fixed

fasmat committed:
VRFSignature is 80 bytes

fasmat committed:
Fix tests

countvonzero commented on beacon/message.go:
```diff
@@ -17,31 +17,35 @@ type ProposalVrfMessage struct {
 	EpochID        types.EpochID
 	RoundID        types.RoundID
-	VotesBitVector []byte
+	VotesBitVector []byte `scale:"max=32"` // TODO(mafa): check if max. value is correct
```
the protocol guarantees that there are very low probability that there will be over 100 proposals.
so a bit vector of 100 bits is 13 bytes. i think 16 byte may suffice.

countvonzero commented on beacon/message.go:
```diff
@@ -17,31 +17,35 @@ type ProposalVrfMessage struct {
+}
+
+type Proposal struct {
+	Value []byte `scale:"max=32"` // TODO(mafa): check if max. value is correct
```
4 byte exactly. (types.BeaconSize)

countvonzero commented on common/types/account.go:
```diff
@@ -14,7 +14,7 @@ type Account struct {
 	Balance         uint64
 	TemplateAddress *Address
-	State           []byte
+	State           []byte `scale:"max=1024"` // TODO(mafa): check if max. value is correct
```
how do you come up with 1024 byte here?

fasmat replied:
I just made up a number, what is a better size?

countvonzero commented on common/types/ballot.go:
```diff
@@ -135,11 +135,11 @@ type Votes struct {
+	Against []Vote `scale:"max=800"` // TODO(mafa): check if this is the right max size
 	// Abstain on layers until they are terminated.
-	Abstain []LayerID
+	Abstain []LayerID `scale:"max=800"` // TODO(mafa): check if this is the right max size
```
can @dshulyak @fasmat tell me the rough math of how these max values are derived?

countvonzero replied:
i mean 800 sounds excessive for VotingEligibility (considering we are aiming total 50 proposals per layer)
for Support/Against/Abstain this means we will not find a reasonable base ballot for 800 layers?

fasmat replied:
whenever I didn't know what to use as max value I just put a value there that I thought was reasonable and added a `TODO`. I can lower the values to `50` each if that is a better maximum value?

countvonzero commented on common/types/ballot.go:
```diff
@@ -197,7 +197,7 @@ func (o *Opinion) MarshalLogObject(encoder log.ObjectEncoder) error {
 type EpochData struct {
 	// from the smesher's view, the set of ATXs eligible to vote and propose block content in this epoch
-	ActiveSet []ATXID
+	ActiveSet []ATXID `scale:"max=800"` // TODO(mafa): check if this is the right max size
```
this mean we will have max 800 ATXs in the active set. this is def. too small

fasmat replied:
What would be a better max. size in your opinion?

countvonzero reviewed (commented):
we should work out the exact max number for each of the protocol msgs instead of leaving them as TODOs at this point.

fasmat committed:
Make Proposal an Array instead of a slice

fasmat committed:
Cleanup

fasmat committed:
Cleanup

fasmat committed:
Fix incorrect size for Vrfsignature

fasmat committed:
More fixes

fasmat committed:
More fixes

fasmat committed:
Merge remote-tracking branch 'origin/develop' into 4095-enforce-max-size-go-scale

countvonzero commented:
@fasmat 
what do you think of making a table that list all the max sizes decision that are made in this PR?
then we can go through them one-by-one on the dev sync meeting.
this would be great protocol documentation as well.

fasmat committed:
Fix failing tests

fasmat committed:
Fix size for active set

fasmat commented:
@countvonzero @dshulyak @poszu 

I added all fields of all types that I added size specifications for to the PR description. I don't think this is a good base for a spec, since all the context for those types are missing, but maybe it helps in finishing this PR by agreeing on the sizes of those fields.

There is also at least one type `type.PoetRequest` that contains itself. So theoretically even with max. sizes it can be infinitely big. Does it have to be recursive?

fasmat committed:
Use errors.As instead of custom function

fasmat committed:
Fix failing test

fasmat committed:
Remove unused import

fasmat committed:
Update poet

fasmat committed:
Fix invalid proofs in tests

fasmat commented:
bors try

countvonzero commented:
> @countvonzero @dshulyak @poszu
> 
> I added all fields of all types that I added size specifications for to the PR description. I don't think this is a good base for a spec, since all the context for those types are missing, but maybe it helps in finishing this PR by agreeing on the sizes of those fields.

the PR itself is a documentation. (not spec). i think it's better to put a table in the PR description that lists all the max size enforced. that way, people just need to read the PR description instead of reading through the changes.

my preference would be using that table to do a team discussion. it's a good opportunity to share the parameters amongst the team members so we are on the same page.

countvonzero commented:
> the PR itself is a documentation. (not spec). i think it's better to put a table in the PR description that lists all the max size enforced. that way, people just need to read the PR description instead of reading through the changes.

just saw that you already did it! ❤️

countvonzero commented:
i updated the table in the PR description and added my thoughts. 
some of the max values should correspond to a config, but should not be exactly the configured value, since we may want to change the configured value without doing a software update

fasmat committed:
Integrate review feedback

fasmat commented:
I updated some of the sizes and the PR description based on feedback from @countvonzero ❤️

fasmat committed:
Fix typos

fasmat committed:
Merge remote-tracking branch 'origin/develop' into 4095-enforce-max-size-go-scale

fasmat committed:
make generate

fasmat committed:
Update poet

fasmat committed:
Update poet

fasmat committed:
Update sizes based on feedback

fasmat committed:
Merge remote-tracking branch 'origin/develop' into 4095-enforce-max-size-go-scale

fasmat committed:
Integrate feedback

fasmat commented:
bors try

countvonzero commented on beacon/message.go:
```diff
@@ -17,31 +19,53 @@ type ProposalVrfMessage struct {
+
+func ProposalFromString(s string) Proposal {
+	var p Proposal
+	copy(p[:], []byte(s))
```
for copy i think you can do just copy(p[:], s) without casting as

countvonzero commented on beacon/message.go:
```diff
@@ -17,31 +19,53 @@ type ProposalVrfMessage struct {
+	return p
+}
+
+func (p Proposal) String() string {
```
i don't think i like this name. this can be easily used in the log.Stringer() interface and the output would be garbage binary.

fasmat replied:
I completely got rid of the conversion of `Proposal` from and to `string` so this method isn't necessary any more.

countvonzero commented on beacon/message.go:
```diff
@@ -17,31 +19,53 @@ type ProposalVrfMessage struct {
 	EpochID                   types.EpochID
-	ValidProposals            [][]byte
-	PotentiallyValidProposals [][]byte
+	ValidProposals            []Proposal `scale:"max=1000"`
```
add comment:

// number of proposals is under 100 with a high probability. 10X for safety margin.

countvonzero commented on common/fixture/transaction_results.go:
```diff
@@ -71,7 +71,8 @@ func (g *TransactionResultGenerator) Next() *types.TransactionWithResult {
 	tx.Layer = g.Layers[g.rng.Intn(len(g.Layers))]
 	if lth := g.rng.Intn(len(g.Addrs)); lth > 0 {
-		tx.Addresses = make([]types.Address, lth)
+		// tx.Addresses will only be 1 or 2 addresses
```
rm

countvonzero commented on common/fixture/transaction_results.go:
```diff
@@ -71,7 +71,8 @@ func (g *TransactionResultGenerator) Next() *types.TransactionWithResult {
 	if lth := g.rng.Intn(len(g.Addrs)); lth > 0 {
-		tx.Addresses = make([]types.Address, lth)
+		// tx.Addresses will only be 1 or 2 addresses
+		tx.Addresses = make([]types.Address, lth%2+1)
```
use 10 instead

countvonzero commented on common/types/activation.go:
```diff
@@ -466,9 +466,9 @@ func (p *PoetProof) MarshalLogObject(encoder log.ObjectEncoder) error {
-	RoundID       string
-	Signature     []byte
+	PoetServiceID []byte `scale:"max=32"` // public key of the PoET service
+	RoundID       string `scale:"max=32"`
```
add TODO to convert to int

countvonzero commented on common/types/ballot.go:
```diff
@@ -35,18 +35,18 @@ func (id *BallotID) DecodeScale(d *scale.Decoder) (int, error) {
 	// Votes field is not signed.
-	Votes Votes
-	// the proof of the smesher's eligibility to vote and propose block content in this epoch.
+	Votes Votes `scale:"max=10000"` // TODO(mafa): check again why this is the right max size
```
you don't need limit here. `Votes` is just a plain struct

countvonzero commented on common/types/ballot.go:
```diff
@@ -35,18 +35,18 @@ func (id *BallotID) DecodeScale(d *scale.Decoder) (int, error) {
+	// the proof of the smeshers eligibility to vote and propose block content in this epoch.
 	// Eligibilities must be produced in the ascending order.
-	EligibilityProofs []VotingEligibility
+	EligibilityProofs []VotingEligibility `scale:"max=500"` // TODO(mafa): check again why this is the right max size
```
the protocol param is ~50 proposals per layer. 10X for safety margin.

countvonzero commented on common/types/ballot.go:
```diff
@@ -135,11 +135,11 @@ type Votes struct {
 	Base BallotID
 	// Support block id at a particular layer and height.
-	Support []Vote
+	Support []Vote `scale:"max=10000"` // TODO(mafa): check again why this is the right max size
```
for Support and Against, the max number of votes would be the tortoise sliding window, which is 10K layers, assuming 1 block per layer.

for Abstain, similar except the vote is one per layer (as opposed to one per block for Support/Against)

countvonzero commented on common/types/transaction_result.go:
```diff
@@ -44,14 +44,13 @@ func (t TransactionStatus) DecodeScale(d *scale.Decoder) (uint8, int, error) {
 type TransactionResult struct {
 	Status  TransactionStatus
-	Message string
+	Message string `scale:"max=1024"` // TODO(mafa): check again why this is the right max size
```
this is just error message for UI/explorer to explain why the transaction failed (if it failed). so it's arbitrary human readable text. 1K is reasonable.

fasmat replied:
added TODO that it should be converted to an error code instead of a message

countvonzero commented on fetch/wire_types.go:
```diff
@@ -11,27 +11,27 @@ import (
 type ResponseMessage struct {
 	Hash types.Hash32
-	Data []byte
+	Data []byte `scale:"max=1000000"` // TODO(mafa): check if this is the right max size
```
i think the max number would be 100K ATXID. s 4MB

countvonzero commented on fetch/wire_types.go:
```diff
@@ -11,27 +11,27 @@ import (
 type RequestBatch struct {
 	ID       types.Hash32
-	Requests []RequestMessage
+	Requests []RequestMessage `scale:"max=20"` // depends on fetch config `BatchSize` which defaults to 20
```
make it 1000 instead. we want to be able to configure this number without release a binary.

countvonzero commented on fetch/wire_types.go:
```diff
@@ -48,22 +48,22 @@ func (r *MeshHashRequest) MarshalLogObject(encoder log.ObjectEncoder) error {
-	Layers []types.LayerID
-	Hashes []types.Hash32
+	Layers []types.LayerID `scale:"max=100"` // same as syncer Config `MaxHashesInReq`
+	Hashes []types.Hash32  `scale:"max=100"` // same as syncer Config `MaxHashesInReq`
```
1K for both too

countvonzero commented on fetch/wire_types.go:
```diff
@@ -11,27 +11,27 @@ import (
 type ResponseBatch struct {
 	ID        types.Hash32
-	Responses []ResponseMessage
+	Responses []ResponseMessage `scale:"max=20"` // depends on fetch config `BatchSize` which defaults to 20
```
1000, same as RequestBatch.Requests

countvonzero commented on fetch/wire_types.go:
```diff
@@ -48,22 +48,22 @@ func (r *MeshHashRequest) MarshalLogObject(encoder log.ObjectEncoder) error {
 
 type MaliciousIDs struct {
-	NodeIDs []types.NodeID
+	NodeIDs []types.NodeID `scale:"max=100000"` // maximum expected number of nodes in the network
```
:s/nodes/ATXs

countvonzero commented on fetch/wire_types.go:
```diff
@@ -48,22 +48,22 @@ func (r *MeshHashRequest) MarshalLogObject(encoder log.ObjectEncoder) error {
-	Ballots []types.BallotID
-	Blocks  []types.BlockID
+	Ballots []types.BallotID `scale:"max=800"`
+	Blocks  []types.BlockID  `scale:"max=800"`
```
this is different from ballots. each layer we expect to have 1 block. so maybe 100 would be a good safety margin.

countvonzero commented on hare/builder.go:
```diff
@@ -68,32 +68,32 @@ func (m *Message) SignedBytes() []byte {
 // Typically used as a collection of commit messages.
 type Certificate struct {
-	Values  []types.ProposalID // the committed set S
+	Values  []types.ProposalID `scale:"max=800"` // the committed set S // TODO(mafa): check if this is the right max size
```
500. this is the same as number of expected proposals per layer with 10X margin

countvonzero commented on hare/builder.go:
```diff
@@ -68,32 +68,32 @@ func (m *Message) SignedBytes() []byte {
 // AggregatedMessages is a collection of messages.
 type AggregatedMessages struct {
-	Messages []Message
+	Messages []Message `scale:"max=800"` // TODO(mafa): check if this is the right max size
```
should be consistent with N with safety margin -> 1000

countvonzero commented on hare/builder.go:
```diff
@@ -68,32 +68,32 @@ func (m *Message) SignedBytes() []byte {
 	Type           MessageType
 	CommittedRound uint32              // the round Values (S) is committed (Ki)
-	Values         []types.ProposalID  // the set S. optional for commit InnerMsg in a certificate
+	Values         []types.ProposalID  `scale:"max=800"` // the set S. optional for commit InnerMsg in a certificate // TODO(mafa): check if this is the right max size
```
same as Certificate.Values

countvonzero commented on hare/flows_test.go:
```diff
@@ -289,21 +290,29 @@ func Test_multipleCPs(t *testing.T) {
+					if outputs[idx] == nil {
+						outputs[idx] = make(map[types.LayerID]LayerOutput)
+					}
+					outputs[idx][out.Layer] = out
```
if you change this, you should also change Test_multipleCPsAndIterations too.

countvonzero reviewed (commented):
for PR description:
- proposalList remove from table
- ballot.Votes remove from table

question:
for pubkey, ed signature and vrf signature, why didn't you go with fixed-size array [32]byte, [64]byte and [80]byte?

fasmat committed:
Merge remote-tracking branch 'origin/develop' into 4095-enforce-max-size-go-scale

fasmat committed:
Avoid converting proposals to and from strings

fasmat committed:
Merge remote-tracking branch 'origin/develop' into 4095-enforce-max-size-go-scale

fasmat committed:
Integrate review feedback

fasmat committed:
Integrate updates from poet

fasmat committed:
Merge remote-tracking branch 'origin/develop' into 4095-enforce-max-size-go-scale

fasmat committed:
Make generate

fasmat commented:
bors try

countvonzero commented on fetch/wire_types.go:
```diff
@@ -48,22 +48,22 @@ func (r *MeshHashRequest) MarshalLogObject(encoder log.ObjectEncoder) error {
 type LayerData struct {
-	Ballots []types.BallotID
-	Blocks  []types.BlockID
+	Ballots []types.BallotID `scale:"max=800"`
```
i'll set this to 500 just to be consistent with number of proposals expected in a layer

countvonzero commented on fetch/wire_types.go:
```diff
@@ -48,22 +48,22 @@ func (r *MeshHashRequest) MarshalLogObject(encoder log.ObjectEncoder) error {
 
 type EpochData struct {
-	AtxIDs []types.ATXID
+	AtxIDs []types.ATXID `scale:"max=1000000"`
```
i think we agreed on 100K, the max atx ids in an epoch.

fasmat replied:
looks like a typo - fixed

countvonzero reviewed (approved):
i updated the table to fill out some missing comments.

fasmat committed:
Merge remote-tracking branch 'origin/develop' into 4095-enforce-max-size-go-scale

fasmat committed:
Integrate feedback

fasmat commented:
bors merge

fasmat committed:
make generate

fasmat commented:
bors merge

fasmat commented:
bors merge
